packages/store/addon/-private/system/references/reference.ts

```diff
@@ -42,8 +49,15 @@ interface Reference {
 }
 abstract class Reference {
   public recordData: InternalModel['_recordData'];
-  constructor(public store: CoreStore, internalModel: InternalModel) {
-    INTERNAL_MODELS.set(this, internalModel);
+  constructor(public store: CoreStore, identifierOrInternalModel: InternalModel | StableRecordIdentifier) {
+    if (RECORD_ARRAY_MANAGER_IDENTIFIERS) {
+      let internalModel = internalModelFactoryFor(this.store).peek(identifierOrInternalModel as StableRecordIdentifier);
+      if (internalModel) {
+        INTERNAL_MODELS.set(this, internalModel);
+      }
+    } else {
+      INTERNAL_MODELS.set(this, identifierOrInternalModel as InternalModel);
```

Comment:
Are we ok coercing variables to a specific type?

Reply:
Yeah, usually a type guard would be great, but we don't have one in this case.